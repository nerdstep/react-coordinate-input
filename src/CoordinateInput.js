// @ts-check
// https://github.com/text-mask/text-mask/issues/887
// https://github.com/text-mask/text-mask/issues/806
// https://github.com/text-mask/text-mask/pull/807
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import MaskedInput from 'react-text-mask'
import createLatLongPipe from './createLatLongPipe'
import createMask from './createMask'
import {
  convertInput,
  dmsToDecimal,
  normalizeInput,
  parseDMS,
  validateDMS,
} from './utils'

const getMaskOptions = ({
  dmsPrecision,
  degreeChar,
  minuteChar,
  secondChar,
  spacerChar,
}) =>
  Object.assign(
    {},
    {
      degree: degreeChar,
      minute: minuteChar,
      second: secondChar,
      space: spacerChar,
      precision: dmsPrecision,
    }
  )

/**
 * @class CoordinateInput
 */
export default class CoordinateInput extends Component {
  static propTypes = {
    className: PropTypes.string,
    degreeChar: PropTypes.string,
    ddPrecision: PropTypes.number,
    dmsPrecision: PropTypes.number,
    guide: PropTypes.bool,
    minuteChar: PropTypes.string,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    placeholder: PropTypes.string,
    placeholderChar: PropTypes.string,
    inputProps: PropTypes.object,
    inputRef: PropTypes.func,
    secondChar: PropTypes.string,
    showMask: PropTypes.bool,
    spacerChar: PropTypes.string,
    value: PropTypes.string,
  }

  static defaultProps = {
    guide: true,
    showMask: false,
    ddPrecision: 6,
    dmsPrecision: 0,
    placeholder: '04° 08′ 15″ N 162° 03′ 42″ E',
    placeholderChar: '_',
    degreeChar: '°',
    minuteChar: '′',
    secondChar: '″',
    spacerChar: ' ',
    onBlur: () => {},
    onChange: () => {},
  }

  constructor(props) {
    super(props)

    const { dmsPrecision, value } = props

    this.state = {
      mask: createMask(getMaskOptions(props)),
      pipe: createLatLongPipe(dmsPrecision),
      value,
    }
  }

  static getDerivedStateFromProps(props) {
    const mask = createMask(getMaskOptions(props))
    const pipe = createLatLongPipe(props.dmsPrecision)
    const value = convertInput(props.value, props.dmsPrecision)

    return { mask, pipe, value }
  }

  handleBlur = () => {
    if (this.inputRef) {
      const { value } = this.inputRef.inputElement
      this.handleChange({ target: { value } })
    }

    this.props.onBlur()
  }

  handleChange = e => {
    const { ddPrecision, onChange } = this.props
    const { value } = e.target
    const dms = normalizeInput(value)
    const valid = validateDMS(dms)

    // Callback if there's an empty value
    if (value.length === 0) {
      onChange(e, { dd: [], dms: '', dmsArray: [] })

      // Otherwise only callback if the value is a valid DMS
    } else if (valid) {
      const dmsArray = parseDMS(dms)

      // @ts-ignore {https://github.com/Microsoft/TypeScript/pull/24897}
      const lat = dmsToDecimal(...dmsArray[0], ddPrecision)
      // @ts-ignore
      const lon = dmsToDecimal(...dmsArray[1], ddPrecision)

      //console.log('handleChange', { value, dmsArray, lat, lon })

      // Callback with the original event and the converted values
      onChange(e, { dd: [lat, lon], dms, dmsArray })
    }
  }

  setRef = ref => {
    if (ref) {
      this.inputRef = ref

      if (typeof this.props.inputRef === 'function') {
        this.props.inputRef(ref)
      }
    }
  }

  render() {
    const { mask, pipe, value } = this.state
    const {
      className,
      guide,
      inputProps,
      name,
      placeholder,
      placeholderChar,
      showMask,
    } = this.props

    return (
      <MaskedInput
        {...inputProps}
        dir="auto"
        guide={guide}
        className={className}
        key={value}
        mask={mask}
        name={name}
        onBlur={this.handleBlur}
        onChange={this.handleChange}
        pipe={pipe}
        placeholder={placeholder}
        placeholderChar={placeholderChar}
        ref={this.setRef}
        showMask={showMask}
        type="text"
        value={value}
      />
    )
  }
}
