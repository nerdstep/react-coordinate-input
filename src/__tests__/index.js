import '@testing-library/jest-dom'
import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import ReactDOM from 'react-dom'
import CoordinateInput from '../'

const testsValues = [
  {
    testName: 'should call the onChange handler with empty results',
    inputName: 'testInput',
    value: ' ',
    result: {
      value: '',
      unmaskedValue: '',
      dd: [],
      dms: [],
    },
  },
  {
    testName: 'should have dmsPrecision 3',
    inputName: 'testInput',
    value: '90 00 00.000 N 180 00 00.000 E',
    props: {
      dmsPrecision: 3,
    },
    result: {
      value: '90° 00′ 00.000″ N 180° 00′ 00.000″ E',
      unmaskedValue: '900000000N1800000000E',
      dd: [90, 180],
      dms: [
        [90, 0, 0, 'N'],
        [180, 0, 0, 'E'],
      ],
    },
  },
  {
    testName: 'should have ddPrecision 8',
    inputName: 'testInput',
    value: '30 09 08 N 038 09 07 E',
    props: {
      ddPrecision: 8,
    },
    result: {
      value: '30° 09′ 08″ N 038° 09′ 07″ E',
      unmaskedValue: '300908N0380907E',
      dd: [30.15222222, 38.15194444],
      dms: [
        [30, 9, 8, 'N'],
        [38, 9, 7, 'E'],
      ],
    },
  },
]

testsValues.forEach(({ testName, inputName, value, props = {}, result }) => {
  it(`${testName}`, () => {
    const event = { target: { value } }
    const onChange = jest.fn((value, obj) => ({ value, ...obj }))

    const { /* debug, */ getByLabelText } = render(
      <CoordinateInput
        {...props}
        name={inputName}
        onChange={onChange}
        value={value}
      />
    )

    // debug()

    const input = getByLabelText(inputName)

    fireEvent.change(input, event)

    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('name', inputName)
    expect(input).toHaveValue(value)
    expect(onChange).toHaveReturnedWith(result)
    expect(onChange).toHaveBeenCalledTimes(1)
  })
})

describe('CoordinateInput', () => {
  it('is truthy', () => {
    expect(CoordinateInput).toBeTruthy()
  })

  it('renders without crashing', () => {
    ReactDOM.render(
      <CoordinateInput onChange={() => {}} />,
      document.createElement('div')
    )
  })
})
