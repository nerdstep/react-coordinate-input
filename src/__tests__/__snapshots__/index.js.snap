// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoordinateInput should call the onChange handler with empty results 1`] = `
<t
  dir="auto"
  guide={true}
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[nNsS\\]/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[eEwW\\]/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="04° 08′ 15″ N 162° 03′ 42″ E"
  placeholderChar="_"
  render={[Function]}
  showMask={false}
  type="text"
/>
`;

exports[`CoordinateInput should call the onChange handler with the converted results 1`] = `
<t
  dir="auto"
  guide={true}
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      ".",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[nNsS\\]/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      ".",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[eEwW\\]/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="04° 08′ 15″ N 162° 03′ 42″ E"
  placeholderChar="_"
  render={[Function]}
  showMask={false}
  type="text"
/>
`;

exports[`CoordinateInput should match the snapshot 1`] = `
<t
  dir="auto"
  guide={true}
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[nNsS\\]/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[eEwW\\]/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="04° 08′ 15″ N 162° 03′ 42″ E"
  placeholderChar="_"
  render={[Function]}
  showMask={false}
  type="text"
/>
`;

exports[`CoordinateInput should match the snapshot 2`] = `
<t
  dir="auto"
  guide={true}
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      ".",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[nNsS\\]/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "°",
      " ",
      /\\\\d/,
      /\\\\d/,
      "′",
      " ",
      /\\\\d/,
      /\\\\d/,
      ".",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "″",
      " ",
      /\\[eEwW\\]/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="04° 08′ 15″ N 162° 03′ 42″ E"
  placeholderChar="_"
  render={[Function]}
  showMask={false}
  type="text"
/>
`;

exports[`CoordinateInput should match the snapshot 3`] = `
<t
  dir="auto"
  guide={true}
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      "D",
      " ",
      /\\\\d/,
      /\\\\d/,
      "M",
      " ",
      /\\\\d/,
      /\\\\d/,
      "S",
      " ",
      /\\[nNsS\\]/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      "D",
      " ",
      /\\\\d/,
      /\\\\d/,
      "M",
      " ",
      /\\\\d/,
      /\\\\d/,
      "S",
      " ",
      /\\[eEwW\\]/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  pipe={[Function]}
  placeholder="04° 08′ 15″ N 162° 03′ 42″ E"
  placeholderChar="_"
  render={[Function]}
  showMask={false}
  type="text"
/>
`;
